### DISCUSSION TITLE
`npm` command to point on local `node_modules/.bin/npm`

### DISCUSSION THREAD

**Discussion by abernier** [npm/feedback#313] [CLI] [answered]
I thought it would have been already the case but it seems not...

If `npm` package is installed as local devDependency, then it would be great if the `npm` command points to that version:

```sh
$ npm --version
6.14.12
$
$ # Inside a project (with package.json)
$ npm init -y
$ npm i npm@7 --save-dev
$ 
$ # BELOW IS EXPECTED RESULT
$ npm --version
7.8.0
```

Ie: `7.8.0` as local npm version installed

It would allow a project to depend on a certain version of npm for development purposes.

What your thoughts about that idea?

**Comment by ljharb** [comment]
That’s not how any binary in the ecosystem works - the version reported should be the one you ran. If you want to run the local one, use npx or an npm run-script.

The way a project depends on a specific version of npm is with the “engines” field.

**Comment by MylesBorins** [comment, marked as answer]
Hey @abernier

Unless we completely changed the way npm worked in the path there would be no way to enforce a shell resolving a specific npm. When you run `npm` your shell is determining the location of the binary via the $PATH environment variable.

One thing you could change on your system would be to put `./node_modules/.bin` at the beginning of your path, but I wouldn't advise this as it could have super strange results. I personally have this [at the end of my path](https://github.com/MylesBorins/dot-files/blob/main/.bash_profile#L11)

As @ljharb mentioned you could use `npm scripts` to do this. For example with the follow package.json the command `npm run version` would return `7.8.0`

```json
{
  "name": "test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "version": "npm --version",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "Myles Borins <mylesborins@github.com>",
  "license": "Apache-2.0",
  "dependencies": {
    "npm": "7.8.0"
  }
}
```

**Reply by abernier** [reply]
Just to clarify,

Probably I'm wrong and this idea is crazy, but I tend to think if I declared `npm` as a devDependency (and maybe only in this case), I should then be able to `npm i` or even `npm publish`...etc with the version of npm I've declared in my `devDependencies`.

`npm` binary could detect a such local `devDependencies.npm` and use `node_modules/.bin/npm` instead of the system's one (for dev purposes).

A usecase would be a monorepo using npm7 workspaces: it would be a way to enforce developpers using the right version of npm (the one declared in `devDependencies`), for dev purposes only.

**Reply by MylesBorins** [reply]
@abernier the challenge here, imho, is that we are talking about an entirely new semantic to how npm resolves the binary which itself would be Semver-Major as it would have experience breaking changes if it were to be rolled out.

If you want to explore this concept further you can open an issue over the the RFC repo https://github.com/npm/rfcs
